sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,695 @@
+                            bit_index)) {
+    *state_out = kDifSpiDeviceIrqStateEnabled;
+  } else {
+    *state_out = kDifSpiDeviceIrqStateDisabled;
+  }
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_clear_all(
+    const dif_spi_device_t *spi) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  // The state register is a write-one-clear register.
+  uint32_t irq_state =
+      mmio_region_read32(spi->base_addr, SPI_DEVICE_INTR_STATE_REG_OFFSET);
+  mmio_region_write32(spi->base_addr, SPI_DEVICE_INTR_STATE_REG_OFFSET,
+                      irq_state);
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_enable(
+    const dif_spi_device_t *spi, dif_spi_device_irq_type_t type,
+    dif_spi_device_irq_state_t state) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t bit_index;
+  switch (type) {
+    case kDifSpiDeviceIrqTypeRxFull:
+      bit_index = SPI_DEVICE_INTR_ENABLE_RXF;
+      break;
+    case kDifSpiDeviceIrqTypeRxAboveLevel:
+      bit_index = SPI_DEVICE_INTR_ENABLE_RXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeTxBelowLevel:
+      bit_index = SPI_DEVICE_INTR_ENABLE_TXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeRxError:
+      bit_index = SPI_DEVICE_INTR_ENABLE_RXERR;
+      break;
+    case kDifSpiDeviceIrqTypeRxOverflow:
+      bit_index = SPI_DEVICE_INTR_ENABLE_RXOVERFLOW;
+      break;
+    case kDifSpiDeviceIrqTypeTxUnderflow:
+      bit_index = SPI_DEVICE_INTR_ENABLE_TXUNDERFLOW;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  switch (state) {
+    case kDifSpiDeviceIrqStateEnabled:
+      mmio_region_nonatomic_set_bit32(
+          spi->base_addr, SPI_DEVICE_INTR_ENABLE_REG_OFFSET, bit_index);
+      break;
+    case kDifSpiDeviceIrqStateDisabled:
+      mmio_region_nonatomic_clear_bit32(
+          spi->base_addr, SPI_DEVICE_INTR_ENABLE_REG_OFFSET, bit_index);
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
+  }
+
+  return kDifSpiDeviceResultOk;
+}
+
+dif_spi_device_result_t dif_spi_device_irq_force(
+    const dif_spi_device_t *spi, dif_spi_device_irq_type_t type) {
+  if (spi == NULL) {
+    return kDifSpiDeviceResultBadArg;
+  }
+
+  uint32_t bit_index;
+  switch (type) {
+    case kDifSpiDeviceIrqTypeRxFull:
+      bit_index = SPI_DEVICE_INTR_TEST_RXF;
+      break;
+    case kDifSpiDeviceIrqTypeRxAboveLevel:
+      bit_index = SPI_DEVICE_INTR_TEST_RXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeTxBelowLevel:
+      bit_index = SPI_DEVICE_INTR_TEST_TXLVL;
+      break;
+    case kDifSpiDeviceIrqTypeRxError:
+      bit_index = SPI_DEVICE_INTR_TEST_RXERR;
+      break;
+    case kDifSpiDeviceIrqTypeRxOverflow:
+      bit_index = SPI_DEVICE_INTR_TEST_RXOVERFLOW;
+      break;
+    case kDifSpiDeviceIrqTypeTxUnderflow:
+      bit_index = SPI_DEVICE_INTR_TEST_TXUNDERFLOW;
+      break;
+    default:
+      return kDifSpiDeviceResultBadArg;
```

Comment:
All enum variants are covered. `default` case is not really needed.

Reply:
Done.